drivers/mtd/gd55.c

```diff
@@ -0,0 +1,2082 @@
+                             uint8_t status)
+{
+  off_t        protstart;
+  off_t        protend;
+  off_t        protsize;
+  unsigned int bp;
+
+  /* the BP field is essentially the power-of-two of the number of 64k
+   * sectors that are protected, saturated to the device size.
+   * The msb determines if protection is:
+   *   - top down  (msb not set)
+   *   - bottom up (msb set)
+   */
+
+  bp = (status & GD55_SR_BP_MASK);
+  bp &= ~GD55_STATUS_TB_MASK; /* Ignore top/bottom for now */
+  bp >>= GD55_SR_BP_SHIFT;
+
+  if (bp == 0)
+    {
+      return false;
+    }
+
+  protsize = GD55_BP_SIZE;
+  protsize <<= (bp - 1);
+  protend = GD55_SECTOR_SIZE * priv->nsectors;
+  if (protsize > protend)
+    {
+      protsize = protend;
+    }
+
+  /* The final protection range then depends on if the protection region is
+   * configured top-down or bottom up.
+   */
+
+  if ((status & GD55_STATUS_TB_BOTTOM))
+    {
+      protstart = 0;
+      protend   = protstart + protsize;
+    }
+  else
+    {
+      protstart = protend - protsize;
+
+      /* protend already computed above */
+    }
+
+  return (addr >= protstart && addr < protend);
+}
+
+#ifdef CONFIG_MTD_GD55_SECTOR512
+/****************************************************************************
+ * Name: gd55_flush_cache
+ *
+ * Description:
+ *   If the cache is dirty (meaning that it no longer matches the old FLASH
+ *   contents) or was erased (with the cache containing the correct FLASH
+ *   contents), then write the cached erase block to FLASH.
+ *
+ * Input Parameters:
+ *   priv       - a reference to the device structure
+ *
+ * Returned Value:
+ *   Success (OK) or fail (negated error code)
+ *
+ ****************************************************************************/
+
+static int gd55_flush_cache(FAR struct gd55_dev_s *priv)
+{
+  int ret = OK;
+
+  if (IS_DIRTY(priv) || IS_ERASED(priv))
+    {
+      off_t address;
+
+      /* Convert the erase sector number into a FLASH address */
+
+      address = (off_t)priv->esectno << GD55_SECTOR_SHIFT;
+
+      /* Write entire erase block to FLASH */
+
+      ret = gd55_write_page(priv, priv->sector, address, GD55_SECTOR_SIZE);
+      if (ret < 0)
+        {
+          ferr("ERROR: gd55_write_page failed: %d\n", ret);
+        }
+
+      /* The cache is no long dirty and the FLASH is no longer erased */
+
+      CLR_DIRTY(priv);
+      CLR_ERASED(priv);
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: gd55_read_cache
+ *
+ * Description:
+ *   Read cached data
+ *
+ * Input Parameters:
+ *   priv       - a reference to the device structure
+ *   sector     = sector to read
+ *
+ * Returned Value:
+ *   Success (OK) or fail (negated error code)
+ *
+ ****************************************************************************/
+
+static FAR uint8_t *gd55_read_cache(FAR struct gd55_dev_s *priv,
+                                       off_t sector)
```

Comment:
align

Reply:
as above - checked others for too: hope I didn't miss any!